enhancements/microshift/microshift-tls-cipher.md

```diff
@@ -0,0 +1,236 @@
+are using TLS. In order to secure the API server only TLS 1.2 or above is
+allowed.
+
+Cipher suites are key to the security of TLS handshakes and information
+exchange, and different versions of TLS use different cipher suites. Some TLS
+versions (like 1.0 and 1.1) have been deprecated in favor of new ones, being
+1.2 the most widespread. As with TLS versions, the cipher suites have also
+evolved and some of them have been compromised and are now unsafe for use.
+For these reasons, the possibility of restricting which cipher suites to use
+becomes important for serving endpoints, as they allow administrators to
+protect their APIs.
+
+TLS 1.3 has had changes in cipher suites to simplify them, meaning the
+available suites has been reduced to only a handful. Most cipher suites from
+TLS 1.2 are incompatible with 1.3. Further details on this topic can be found
+in the proposal section.
+
+MicroShift API server is using TLS 1.2 as the minimum version, supporting also
+1.3. Cluster admins who are using tools with TLS 1.2 or above may need to
+comply with security policies restricting which cipher suites can be used.
+
+### User Stories
+As a MicroShift admin, I want to be able to configure TLS cipher suites for the
+API endpoints.
+
+As a MicroShift admin, I want to be able to configure the minimum TLS version
+for the API endpoints.
+
+### Goals
+* Allow users to configure which TLS version and which TLS ciphers are used in
+  API endpoints of MicroShift.
+
+### Non-Goals
+N/A
+
+## Proposal
+Introduce configuration options for API server cipher suites.
+```yaml
+apiServer:
+  tls:
+    # Defaults to the suites of the configured minVersion. When using TLS 1.3
+    # the cipherSuites user input is ignored and replaced with TLS 1.3 default
+    # suites.
+    cipherSuites:
+    - <cipher suite 1>
+    - ...
+    minVersion: <VersionTLS12|VersionTLS13> # Defaults to 1.2
+```
+The API server will use the configured minimum TLS version (which defaults to
+1.2) and the cipher suites. Note that cipher suites may be unconfigured, in
+which case it will automatically pick up the defaults for the configured
+minimum version.
+
+The default cipher suites for 1.2 are:
+* TLS_AES_128_GCM_SHA256
+* TLS_AES_256_GCM_SHA384
+* TLS_CHACHA20_POLY1305_SHA256
+* TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+* TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+* TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+* TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+* TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+* TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+
+The default cipher suites for 1.3 are:
+* TLS_AES_128_GCM_SHA256
+* TLS_AES_256_GCM_SHA384
+* TLS_CHACHA20_POLY1305_SHA256
+
+For simplicity, the API server cipher suites will also apply to internal
+control plane components:
+* API server
+* Kubelet
+* Kube controller manager
+* Kube scheduler
+* Etcd
+* Route controller manager
+
+Except for API server the rest of the components are internal to MicroShift
+and require secure connections too.
```

Comment:
Seems like something that should remain under the control of the microshift team, rather than assuming user settings. In case a suite becomes deprecated, it'd be preferable allow the team to react with a z-stream patch and notify users to update.

Reply:
Good point. When I was writing the enhancement I was picturing a use case where customers have restrictions on what to use even internally (imagine some system where audit policies forbid specific suites).
If internal components cipher suites were to be configurable the file would be a bit convoluted.
If there is no such use case then I agree with you that we could just handle it internally (like defaulting to latest version, TLS 1.3 and its default suites).